### PR TITLE
Migrate project to xlibb

### PR DESCRIPTION
**Pull request by Ishad-M-I-M**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+2. To generate the gateway executable run the following command in the terminal.
+
+```bash
+./gateway.sh -s <supergraphPath> -p <port> -o <outputPath>
+```
+>**Note:**\
+> `supergraphPath` is a mandatory argument. \
+> If the port is not provided the default port 9090 will be used. \
+> If the output path is not provided the gateway executable will be generated in the current directory. 
+3. Run the generated gateway executable `jar` with
```
Comment (ThisaruGuruge) [inline README.md]: This should ideally be a part of the step. Not an additional note.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+```
+or
+```bash
+java -jar <executable-file>
+```
+
+### Try out the example
+1. Navigate into `examples/astronauts_missions_example` directory. There's two federated graphql services and a supergraph schema.
+2. In terminal execute `./gateway.sh -s supergraph.graphql` to generate the gateway executable. It'll generate the executable named as `supergraph-gateway.jar` in the current directory.
+3. In terminal execute `bal run` inside both `astronauts_service` and `missions_service` directories to start the subgraph services.
```
Comment (ThisaruGuruge) [inline README.md:38]: Let's split this up and create separate steps to run two subgraphs.

File: native/src/main/java/io/xlibb/gateway/generator/GatewayQueryPlanGenerator.java [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+                try {
+                    String graph = getGraphOfJoinTypeArgument(directive);
+                    String key = getKeyOfJoinTypeArgument(name, directive);
+                    keys.put(graph, key);
+                } catch (GatewayGenerationException | ValidationException ignored) {
+
+                }
+            }
+        }
+
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/generator/GatewayQueryPlanGenerator.java]: Extra empty lines? Shall we check all the places and remove them?

File: native/src/main/java/io/xlibb/gateway/generator/GatewayQueryPlanGenerator.java [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    private void addClientConstantDeclarations(List<ModuleMemberDeclarationNode> nodeList) throws ValidationException {
+        for (Map.Entry<String, JoinGraph> entry :
+                getJoinGraphs(this.graphQLSchema).entrySet()) {
+            nodeList.add(NodeParser.parseModuleMemberDeclaration(
+                    CLIENT_NAME_DECLARATION.replace(CLIENT_NAME_PLACEHOLDER, entry.getKey())
+                            .replace(CLIENT_NAME_VALUE_PLACEHOLDER, entry.getValue().getName())
+            ));
+        }
+    }
+
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/generator/GatewayQueryPlanGenerator.java]:
```suggestion
```

File: native/src/main/java/io/xlibb/gateway/generator/common/CommonUtils.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                if (graph != null && (external == null || !external)) {
+                    return graph;
+                }
+            }
+        }
+
+        if (joinTypeDirectivesOnParent.size() == 1) {
+            for (GraphQLAppliedDirectiveArgument argument : joinTypeDirectivesOnParent.get(0).getArguments()) {
+                Object value = argument.getArgumentValue().getValue();
+                if (argument.getName().equals("graph") && value instanceof EnumValue) {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/generator/common/CommonUtils.java]: Better to define constants for these values (`"graph"`, etc.)

File: native/src/main/java/io/xlibb/gateway/generator/common/Constants.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.xlibb.gateway.generator.common;
+
+/**
+ * Class to store constants used in gateway generation.
+ */
+public class Constants {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/generator/common/Constants.java]: Defining a separate class for constants is considered an [anti-pattern](https://stackoverflow.com/a/15056462/3615862). Refer to [this article](https://www.baeldung.com/java-constants-good-practices) as well. Instead, best practice is to define the constants within the class itself. (Or for commonly used constants, we can use a utility class). 

Of course, there can be debates around whether this is okay, but as a conversion, let's not create a separate constants class.

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Class to hold the utility methods.
+ * */
+public class Utils {
+
+    private static final String[] KEYWORDS = new String[] {"abort", "aborted", "abstract", "all", "annotation",
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/graphql/Utils.java:43]: This isn't scalable. We cannot keep track of every Ballerina keyword here. We should find a better way to check for keywords. Probably there is a compiler API to do this.

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    public static final String BALLERINA_STRING_TYPE = "string";
+    public static final String BALLERINA_INT_TYPE = "int";
+    public static final String BALLERINA_FLOAT_TYPE = "float";
+    public static final String BALLERINA_BOOLEAN_TYPE = "boolean";
+    public static final String BALLERINA_ANYDATA_TYPE = "anydata";
+
+    static {
+        BAL_KEYWORDS = Collections.unmodifiableList(Arrays.asList(KEYWORDS));
+        BAL_TYPES = Collections.unmodifiableList(Arrays.asList(TYPES));
+    }
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/xlibb/gateway/graphql/Utils.java]: What's the reason for defining these in a separate static block?

File: native/src/main/java/module-info.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    requires io.ballerina.parser;
+    requires com.graphqljava;
+    requires com.graphqljava.extendedscalars;
+    requires java.net.http;
+    requires io.ballerina.tools.api;
+    requires io.ballerina.runtime;
+    requires org.apache.commons.io;
+    requires io.ballerina.formatter.core;
+    requires org.slf4j;
+}
```
Comment (ThisaruGuruge) [inline native/src/main/java/module-info.java]:
```suggestion
}

```

File: native/src/test/java/io/xlibb/gateway/generator/common/CommonUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    @DataProvider(name = "SchemaAndQueryTypesProvider")
+    public Object[][] getSchemaAndQueryTypes() throws ValidationException, IOException {
+        return new Object[][]{
+                {GatewayTestUtils.getGatewayProject("two_entities", tmpDir), new String[]{
+                        "astronaut", "astronauts", "mission", "missions", "serviceName", "isExist"
+                }}
+        };
+    }
+
+    @Test(description = "Test get mutation types", dataProvider = "SchemaAndMutationTypesProvider")
```
Comment (ThisaruGuruge) [inline native/src/test/java/io/xlibb/gateway/generator/common/CommonUtilTest.java]: I don't think these descriptions are adding any value. The test function name is self-explanatory.

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java (hunk 7 above)
Comment (Ishad-M-I-M) [reply native/src/main/java/io/xlibb/gateway/graphql/Utils.java:43]: Actually, I fetched these from `graphql-tools` (https://github.com/ballerina-platform/graphql-tools/blob/main/graphql-cli/src/main/java/io/ballerina/graphql/generator/CodeGeneratorConstants.java#L102). 
I'll refactor this. Can I get reference to that compiler API

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java (hunk 8 above)
Comment (Ishad-M-I-M) [reply native/src/main/java/io/xlibb/gateway/graphql/Utils.java]: Same
https://github.com/ballerina-platform/graphql-tools/blob/94063368922edada69f76aed2adb86a4236fa687/graphql-cli/src/main/java/io/ballerina/graphql/generator/CodeGeneratorConstants.java#L123

I'll revisit the code and refactor it leisurely.

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java (hunk 7 above)
Comment (ThisaruGuruge) [reply native/src/main/java/io/xlibb/gateway/graphql/Utils.java:43]: I see. Then let's stick to this for now.

File: native/src/main/java/io/xlibb/gateway/graphql/Utils.java (hunk 8 above)
Comment (ThisaruGuruge) [reply native/src/main/java/io/xlibb/gateway/graphql/Utils.java]: No worries.

File: native/src/main/java/io/xlibb/gateway/generator/common/Constants.java (hunk 6 above)
Comment (Ishad-M-I-M) [reply native/src/main/java/io/xlibb/gateway/generator/common/Constants.java]: Done!

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**